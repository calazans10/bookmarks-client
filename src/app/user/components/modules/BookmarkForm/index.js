import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Form, Field } from 'react-final-form';
import MainForm from '../../../../core/components/modules/MainForm';
import FormGroup from '../../../../core/components/modules/FormGroup';
import ButtonSubmit from '../../../../core/components/modules/ButtonSubmit';
import { required } from '../../../../../utils/validators';

const BookmarkForm = ({ legend, action, title, url, onSubmit }) => (
  <Form
    initialValues={{ title, url }}
    onSubmit={onSubmit}
    render={({ handleSubmit }) => (
      <MainForm legend={legend} onSubmit={handleSubmit}>
        <Field
          name="title"
          validate={required}
          render={({ input, meta }) => (
            <FormGroup input={input} meta={meta} label="Title" type="text" />
          )}
        />
        <Field
          name="url"
          validate={required}
          render={({ input, meta }) => (
            <FormGroup input={input} meta={meta} label="Url" type="url" />
          )}
        />
        <div>
          <ButtonSubmit>{action}</ButtonSubmit>
          <Link to="/bookmarks">Cancel</Link>
        </div>
      </MainForm>
    )}
  />
);

BookmarkForm.propTypes = {
  legend: PropTypes.string.isRequired,
  action: PropTypes.string.isRequired,
  title: PropTypes.string,
  url: PropTypes.string,
  onSubmit: PropTypes.func.isRequired,
};

BookmarkForm.defaultProps = {
  title: '',
  url: '',
};

export default BookmarkForm;
