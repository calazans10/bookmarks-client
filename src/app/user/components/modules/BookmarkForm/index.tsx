import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers';
import * as yup from 'yup';
import { BookmarkData } from 'app/user/types';
import MainForm from 'app/core/components/modules/MainForm';
import FormGroup from 'app/core/components/modules/FormGroup';
import ButtonSubmit from 'app/core/components/modules/ButtonSubmit';

type BookmarkFormProps = {
  legend: string;
  action: string;
  title: string;
  url: string;
  onSubmit: (values: BookmarkData) => Promise<void>;
} & typeof defaultProps;

const defaultProps = {
  title: '',
  url: '',
};

const schema = yup.object().shape({
  title: yup.string().required(),
  url: yup.string().url().required(),
});

const BookmarkForm = ({ legend, action, title, url, onSubmit }: BookmarkFormProps) => {
  const { register, handleSubmit, errors } = useForm<BookmarkData>({
    resolver: yupResolver(schema),
  });

  return (
    <MainForm legend={legend} onSubmit={handleSubmit(onSubmit)}>
      <FormGroup
        name="title"
        label="Title"
        type="text"
        defaultValue={title}
        inputRef={register}
        error={errors.title?.message}
      />
      <FormGroup
        name="url"
        label="Url"
        type="url"
        defaultValue={url}
        inputRef={register}
        error={errors.url?.message}
      />
      <div>
        <ButtonSubmit>{action}</ButtonSubmit>
        <Link to="/bookmarks">Cancel</Link>
      </div>
    </MainForm>
  );
};

BookmarkForm.defaultProps = defaultProps;

export default BookmarkForm;
