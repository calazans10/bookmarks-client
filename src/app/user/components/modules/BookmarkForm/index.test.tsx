import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import BookmarkForm from './';

describe('BookmarkForm', () => {
  it('renders without crashing', () => {
    const { container } = render(
      <BookmarkForm legend="A legend" action="Save" onSubmit={jest.fn()} />,
      { wrapper: MemoryRouter }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
