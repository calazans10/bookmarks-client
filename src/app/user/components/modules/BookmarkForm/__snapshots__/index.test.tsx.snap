// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkForm renders without crashing 1`] = `
<form
  autocomplete="off"
  class="sc-AxjAm hYYXIy"
  novalidate=""
>
  <fieldset>
    <legend
      class="sc-AxirZ dBtpcc"
    >
      A legend
    </legend>
    <div
      class="sc-AxiKw bLDZyO"
    >
      <label
        class="sc-AxhCb ieTFEo"
        for="title"
      >
        Title
      </label>
      <input
        class="sc-AxhUy ePtFHq"
        id="title"
        name="title"
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-AxiKw bLDZyO"
    >
      <label
        class="sc-AxhCb ieTFEo"
        for="url"
      >
        Url
      </label>
      <input
        class="sc-AxhUy ePtFHq"
        id="url"
        name="url"
        placeholder="Url"
        type="url"
        value=""
      />
    </div>
    <div>
      <button
        class="sc-AxheI dSQsuG"
        type="submit"
      >
        Save
      </button>
      <a
        href="/bookmarks"
      >
        Cancel
      </a>
    </div>
  </fieldset>
</form>
`;
